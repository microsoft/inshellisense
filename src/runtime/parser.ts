// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

export type CommandToken = {
  token: string;
  complete: boolean;
  isOption: boolean;
  isPersistent?: boolean;
  isPath?: boolean;
  isPathComplete?: boolean;
};

const cmdDelim = /(\|\|)|(&&)|(;)|(\|)/;
const spaceRegex = /\s/;

export const parseCommand = (command: string): CommandToken[] => {
  const lastCommand = command.split(cmdDelim).at(-1)?.trimStart();
  return lastCommand ? lex(lastCommand) : [];
};

const lex = (command: string): CommandToken[] => {
  const tokens: CommandToken[] = [];
  let [readingQuotedString, readingFlag, readingCmd] = [false, false, false];
  let readingIdx = 0;
  let readingQuoteChar = "";

  [...command].forEach((char, idx) => {
    const reading = readingQuotedString || readingFlag || readingCmd;
    if (!reading && (char === `'` || char === `"`)) {
      [readingQuotedString, readingIdx, readingQuoteChar] = [true, idx, char];
      return;
    } else if (!reading && char === `-`) {
      [readingFlag, readingIdx] = [true, idx];
      return;
    } else if (!reading && !spaceRegex.test(char)) {
      [readingCmd, readingIdx] = [true, idx];
      return;
    }

    if (readingQuotedString && char === readingQuoteChar && command.at(idx - 1) !== "\\") {
      readingQuotedString = false;
      const complete = idx + 1 < command.length && spaceRegex.test(command[idx + 1]);
      tokens.push({
        token: command.slice(readingIdx, idx + 1),
        complete,
        isOption: false,
      });
    } else if ((readingFlag && spaceRegex.test(char)) || char === "=") {
      readingFlag = false;
      tokens.push({
        token: command.slice(readingIdx, idx),
        complete: true,
        isOption: true,
      });
    } else if (readingCmd && spaceRegex.test(char)) {
      readingCmd = false;
      tokens.push({
        token: command.slice(readingIdx, idx),
        complete: true,
        isOption: false,
      });
    }
  });

  const reading = readingQuotedString || readingFlag || readingCmd;
  if (reading) {
    tokens.push({
      token: command.slice(readingIdx),
      complete: false,
      isOption: readingFlag,
    });
  }

  return tokens;
};
