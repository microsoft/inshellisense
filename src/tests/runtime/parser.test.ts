// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { parseCommand } from "../../runtime/parser.js";

const testData = [
  { command: `cmd --flag value` },
  { command: `cmd --flag=value` },
  { command: `cmd --flag='value' ` },
  { command: `cmd --flag="value" ` },
  { command: `cmd 'value' ` },
  { command: `cmd value ` },
  { command: `cmd -f` },
  { command: `cmd -f=value ` },
  { command: `cmd -f value ` },
  { command: `cmd -f 'value' ` },
  { command: `cmd -f="value" ` },
  { command: `cmd -f='val` },
  { command: `cmd -f` },
  { command: `cmd -f=` },
  { command: `cmd -f ` },
  { command: `cmd` },
  { command: `cmd ` },
  { command: `cmd "value' ` },
  { command: `cmd "value'\\"\\"" ` },
  { command: `cmd1 | cmd2 ` },
  { command: `cmd1 -` },
];

describe(`parseCommand`, () => {
  testData.forEach(({ command }) => {
    test(command, () => {
      expect(parseCommand(command)).toMatchSnapshot();
    });
  });
});
