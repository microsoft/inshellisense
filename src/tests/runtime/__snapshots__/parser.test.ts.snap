// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseCommand cmd  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
]
`;

exports[`parseCommand cmd "value'  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": false,
    "isOption": false,
    "token": ""value' ",
  },
]
`;

exports[`parseCommand cmd "value'\\"\\""  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": false,
    "token": ""value'\\"\\""",
  },
]
`;

exports[`parseCommand cmd 'value'  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "'value'",
  },
]
`;

exports[`parseCommand cmd --flag value 1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "--flag",
  },
  {
    "complete": false,
    "isOption": false,
    "token": "value",
  },
]
`;

exports[`parseCommand cmd --flag="value"  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "--flag",
  },
  {
    "complete": true,
    "isOption": false,
    "token": ""value"",
  },
]
`;

exports[`parseCommand cmd --flag='value'  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "--flag",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "'value'",
  },
]
`;

exports[`parseCommand cmd --flag=value 1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "--flag",
  },
  {
    "complete": false,
    "isOption": false,
    "token": "value",
  },
]
`;

exports[`parseCommand cmd -f  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "-f",
  },
]
`;

exports[`parseCommand cmd -f 'value'  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "-f",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "'value'",
  },
]
`;

exports[`parseCommand cmd -f 1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": false,
    "isOption": true,
    "token": "-f",
  },
]
`;

exports[`parseCommand cmd -f 2`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": false,
    "isOption": true,
    "token": "-f",
  },
]
`;

exports[`parseCommand cmd -f value  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "-f",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "value",
  },
]
`;

exports[`parseCommand cmd -f= 1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "-f",
  },
]
`;

exports[`parseCommand cmd -f="value"  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "-f",
  },
  {
    "complete": true,
    "isOption": false,
    "token": ""value"",
  },
]
`;

exports[`parseCommand cmd -f='val 1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "-f",
  },
  {
    "complete": false,
    "isOption": false,
    "token": "'val",
  },
]
`;

exports[`parseCommand cmd -f=value  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "-f",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "value",
  },
]
`;

exports[`parseCommand cmd 1`] = `
[
  {
    "complete": false,
    "isOption": false,
    "token": "cmd",
  },
]
`;

exports[`parseCommand cmd value  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "value",
  },
]
`;

exports[`parseCommand cmd1 | cmd2  1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd2",
  },
]
`;

exports[`parseCommand cmd1 - 1`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "cmd1",
  },
  {
    "complete": false,
    "isOption": true,
    "token": "-",
  },
]
`;
